src/GitVersion.LibGit2Sharp/Git/CommitCollection.cs

```diff
@@ -6,11 +6,16 @@ namespace GitVersion.Git;
 internal sealed class CommitCollection : ICommitCollection
 {
     private readonly ICommitLog innerCollection;
+    private readonly Lazy<IReadOnlyCollection<ICommit>> commits;
 
-    internal CommitCollection(ICommitLog collection) => this.innerCollection = collection.NotNull();
+    internal CommitCollection(ICommitLog collection)
+    {
+        this.innerCollection = collection.NotNull();
```

Comment:
Do we need `this.innerCollection` at all, or is it enough as a captured reference inside the Lazy lambda? If it can be removed, it would improve readability and avoid accidental references to `this.innerCollection`.

Reply:
`QueryBy` still uses the inner collection, so we need both currenttly.

Reply:
it's used here https://github.com/GitTools/GitVersion/blob/825f62ca5ad06532d671178c8a683aa200ffc659/src/GitVersion.LibGit2Sharp/Git/CommitCollection.cs#L25-L40, and that's the challenging part, If we can filter on the cached collection the same way git does filtering then we're good

Reply:
Right, so this is an impedance mismatch between the lazy collection and LibGit2's `QueryBy` method? Also, trying to abstract it might put us well into [Hyrum's law](https://www.hyrumslaw.com/) territory. Although we could probably simplify the abstraction if we manage to reduce all of GitVersion's usage to a least common denominator. That's out of scope for this PR, though.